### PR TITLE
gh-91912: Long object free list for medium value

### PR DESCRIPTION
**Pull request by penguin-wwy**
https://github.com/python/cpython/issues/91912

* Issue: gh-91912

### CODE REVIEW

File: Objects/longobject.c [hunk 1]
```diff
@@ -196,11 +224,26 @@ _PyLong_FromMedium(sdigit x)
...
+#ifdef Py_DEBUG
+        assert(state->numfree != -1);
+#endif
+        state->free_list = (PyLongObject *) Py_TYPE(v);
+        state->numfree--;
+    }
+    else
+#endif
+    {
+        /* We could use a freelist here */
```
Comment (MojoVampire) [inline Objects/longobject.c]: If we're adding a freelist, maybe remove this comment.

File: Objects/longobject.c [hunk 2]
```diff
@@ -132,6 +132,15 @@ long_normalize(PyLongObject *v)
     return v;
 }
 
+#if PyLong_MAXFREELIST > 0
+static struct _Py_long_state *
```
Comment (MojoVampire) [inline Objects/longobject.c]: Given how little is done here, might be worth declaring `inline` to ensure no function call overhead paid for such cheap work.

File: Objects/longobject.c [hunk 3]
```diff
@@ -40,7 +40,7 @@ static inline void
 _Py_DECREF_INT(PyLongObject *op)
 {
     assert(PyLong_CheckExact(op));
-    _Py_DECREF_SPECIALIZED((PyObject *)op, (destructor)PyObject_Free);
```
Comment (MojoVampire) [inline Objects/longobject.c:49]: See summary; this is where we used to use `PyObject_Free`.

File: Objects/longobject.c [hunk 4]
```diff
@@ -3051,6 +3094,47 @@ PyLong_AsDouble(PyObject *v)
...
 /* Methods */
 
+void
+_PyLong_ExactDealloc(PyObject *obj)
+{
+    assert(PyLong_CheckExact(obj));
+    PyLongObject *op = (PyLongObject *)obj;
+#if PyLong_MAXFREELIST > 0
+    if (!IS_MEDIUM_VALUE(op)) {
+        PyObject_Del(op);
```
Comment (MojoVampire) [inline Objects/longobject.c]: See summary; where we used to call `PyObject_Free` unconditionally, the conditional call changed to `PyObject_Del`, which is almost certainly incorrect. Same goes for all other uses of `PyObject_Del` here.

File: Objects/longobject.c [hunk 5]
```diff
@@ -3051,6 +3094,47 @@ PyLong_AsDouble(PyObject *v)
...
+{
+    assert(PyLong_Check(op));
+#if PyLong_MAXFREELIST > 0
+    if (PyLong_CheckExact(op)) {
+        _PyLong_ExactDealloc(op);
+    }
+    else
+#endif
+    {
+        Py_TYPE(op)->tp_free(op);
```
Comment (MojoVampire) [inline Objects/longobject.c:3367]: This feels like it should maybe be calling the type's `tp_dealloc`, not `tp_free`? Not going to swear to this, but it feels off to me (then again, calling their `tp_dealloc` could recurse)... Don't take this comment as gospel, I'm probably wrong here.

**Review by MojoVampire — CHANGES_REQUESTED** [review]
In the process of refactoring, what used to be `PyObject_Free` calls (matching the `PyObject_Malloc` calls used to allocate in the first place) became `PyObject_Del` calls (which pair with `PyObject_New`, which is not used here). I don't know whether the two happen to be compatible by coincidence, but even if they are, the calls should still pair with the other calls in the family; `New` with `Del`, `Malloc` with `Free`, while this patch now pairs `Malloc` with `Del`.

File: Objects/longobject.c (hunk 5 above)
Comment (penguin-wwy) [reply Objects/longobject.c:3367]: The argument passed in should be long or a subclass of long, continued use of `tp_dealloc` has a high probability of leading to unlimited recursive calls.

**Review by iritkatriel — CHANGES_REQUESTED** [review]
This has merge conflicts now.

**Comment by penguin-wwy** [comment]
@iritkatriel Resolved conflicts with the main branch.
I have made the requested changes; please review again

**Pull request closed without merging**